test/pinnedLinks.js

```diff
@@ -0,0 +1,148 @@
+    return gUserDatabase.init().then(pinnedLinks => {
+      gPinnedLinks.setLinks(pinnedLinks);
+      assert.lengthOf(gPinnedLinks.links, 0);
+
+      gPinnedLinks.pin(firstLink, 2);
+      return gUserDatabase.load("prefs", "pinnedLinks").then(pinnedLinks => {
+        var storedPinnedLinks = JSON.stringify([null, null, firstLink]);
+        assert.equal(pinnedLinks, storedPinnedLinks);
+        assert.equal(JSON.stringify(gPinnedLinks.links), storedPinnedLinks);
+      });
+    });
+  });
+
+  it("Unpinning a link updates the database and links", () => {
+    return gUserDatabase.init().then(pinnedLinks => {
+      gPinnedLinks.setLinks(pinnedLinks);
+      assert.lengthOf(gPinnedLinks.links, 0);
+
+      gPinnedLinks.pin(firstLink, 2);
+      gPinnedLinks.pin(secondLink, 1);
+      assert.isTrue(gPinnedLinks.isPinned(firstLink));
+      assert.isTrue(gPinnedLinks.isPinned(secondLink));
+
+      gPinnedLinks.unpin(firstLink);
+      assert.isFalse(gPinnedLinks.isPinned(firstLink));
+
+      gPinnedLinks.pin(firstLink, 5);
+
+      return gUserDatabase.load("prefs", "pinnedLinks").then(pinnedLinks => {
+        var storedPinnedLinks = JSON.stringify([null, secondLink, null, null, null, firstLink]);
+        assert.equal(pinnedLinks, storedPinnedLinks);
+        assert.equal(JSON.stringify(gPinnedLinks.links), storedPinnedLinks);
+      });
+    });
+  });
+
+  it("Pinning a directory link turns it to history", () => {
+    return gUserDatabase.init().then(pinnedLinks => {
+      gPinnedLinks.setLinks(pinnedLinks);
+      assert.lengthOf(gPinnedLinks.links, 0);
+
+      gPinnedLinks.pin(directoryLink, 2);
+      return gUserDatabase.load("prefs", "pinnedLinks").then(pinnedLinks => {
+        var storedPinnedLinks = JSON.stringify([null, null, directoryLink]);
+        assert.equal(pinnedLinks, storedPinnedLinks);
+        assert.equal(JSON.stringify(gPinnedLinks.links), storedPinnedLinks);
+        assert.equal(directoryLink.type, "history");
+      });
+    });
+  });
+
+  it("Test replacing a pinned link with another (Used for ended campaigns)", () => {
+    return gUserDatabase.init().then(pinnedLinks => {
+      gPinnedLinks.setLinks(pinnedLinks);
+
+      // Attempting to replace a link that isn't pinned does nothing.
+      gPinnedLinks.replace("http://example0.com/", secondLink);
+      assert.lengthOf(gPinnedLinks.links, 0);
+
+      // Pin firstLink
+      gPinnedLinks.pin(firstLink, 2);
+      assert.equal(gPinnedLinks.links[2], firstLink);
+
+      // Replace the pinned firstLink with secondLink
+      gPinnedLinks.replace(firstLink.url, secondLink);
+      assert.equal(gPinnedLinks.links[2], secondLink);
+    });
+  });
+
+  it("Handling of open database errors", () => {
+    return gUserDatabase.init(gMockObject).then(() => {
+      throw new Error("There should be an error");
+    }, val => {
+      assert.equal(val, "Error: Error during open");
+    });
+  });
+
+  it("Handling of save database errors", () => {
```

Comment:
When I tested the code above, the test framework didn't seem happy with `yield` for some reason. This works though:

```
gUserDatabase.init(["pinnedLinks"]).then(() => {
  var savePromise = gUserDatabase.save("prefs", "pinnedLinks", null, gMockObject);
  return savePromise.should.be.rejected;
});
```